webhook/resources/volume/validator.go

```diff
@@ -156,6 +164,53 @@ func (v *volumeValidator) Update(request *admission.Request, oldObj runtime.Obje
 	return nil
 }
 
+func (v *volumeValidator) validateExpansionSize(oldVolume *longhorn.Volume, newVolume *longhorn.Volume) error {
+	oldSize := oldVolume.Spec.Size
+	newSize := newVolume.Spec.Size
+	if newSize == oldSize {
+		return nil
+	}
+	if newSize < oldSize && !newVolume.Status.ExpansionRequired {
+		return fmt.Errorf("shrinking volume %v size from %v to %v is not supported", newVolume.Name, oldSize, newSize)
+	}
+
+	newKubernetesStatus := &newVolume.Status.KubernetesStatus
+	namespace := newKubernetesStatus.Namespace
+	pvcName := newKubernetesStatus.PVCName
+	if pvcName == "" || newKubernetesStatus.LastPVCRefAt != "" {
+		return nil
+	}
+
+	pvc, err := v.ds.GetPersistentVolumeClaim(namespace, pvcName)
+	if err != nil {
+		return err
+	}
+
+	pvcSCName := *pvc.Spec.StorageClassName
+	pvcStorageClass, err := v.ds.GetStorageClassRO(pvcSCName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "cannot find storage class %v", pvcSCName)
+		}
+		return err
+	}
+	if pvcStorageClass.AllowVolumeExpansion != nil && !*pvcStorageClass.AllowVolumeExpansion {
+		return fmt.Errorf("storage class %v of pvc %v does not allow the volume expansion", pvcSCName, pvc.Name)
+	}
+
+	pvcSpecValue, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+	if !ok {
+		return fmt.Errorf("cannot get request storage of pvc %v", pvc.Name)
+	}
+
+	requestedSize := resource.MustParse(strconv.FormatInt(newSize, 10))
+	if pvcSpecValue.Cmp(requestedSize) < 0 {
+		return fmt.Errorf("waiting for pvc %v expansion from %v to %v", pvcName, pvcSpecValue.Value(), requestedSize.Value())
+	}
```

Comment:
Just revisited.

What does this mean actually? What cases will trigger this? updating volume directly instead of PVC?

The message should be `PVC %v can't be shrinked from %v to %v`. Isn't it?

Reply:
This means PVC size is smaller than requested size now and we need to wait for the PVC expansion first then we can update the volume size to requested size.
When users update volume size of `Volume CR` directly, this will be triggered.

Reply:
Modified the error log. The PVC size needs to be expanded first.